In Astro SSG, scoped slots are breaking the build

### Vue version

3.4.3

### Link to minimal reproduction

https://codesandbox.io/p/devbox/vue-astro-3-4-4-9qmcdd

### Steps to reproduce

Open the code sandbox and let run the `build` task

### What is expected?

I believe it should compile properly.

### What is actually happening?

It is returning the following message:

```
20:08:23 [ERROR] node.scopeIds.forEach is not a function
  Stack trace:
    at Object.enter (/workspaces/sandbox/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:1520:25)
```

### System Info

```shell
System:
    OS: macOS 14.2.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 624.19 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.17.1 - ~/Library/Caches/fnm_multishells/54301_1704309312457/bin/node
    Yarn: 1.22.21 - ~/Library/Caches/fnm_multishells/54301_1704309312457/bin/yarn
    npm: 9.6.7 - ~/Library/Caches/fnm_multishells/54301_1704309312457/bin/npm
    pnpm: 8.14.0 - ~/Library/Caches/fnm_multishells/54301_1704309312457/bin/pnpm
  Browsers:
    Chrome: 120.0.6099.129
    Safari: 17.2.1
  npmPackages:
    vue: ^3.4.4 => 3.4.4
```


### Any additional comments?

It seems to be linked to #9867.
It does not fail on Stackblitz.
I can continue digging into it.

--- a/packages/compiler-sfc/__tests__/compileTemplate.spec.ts
+++ b/packages/compiler-sfc/__tests__/compileTemplate.spec.ts
@@ -441,14 +441,15 @@ test('prefixing edge case for reused AST ssr mode', () => {
   const { descriptor } = parse(src)
   // compileScript triggers importUsageCheck
   compileScript(descriptor, { id: 'xxx' })
-  const { code } = compileTemplate({
-    id: 'xxx',
-    filename: 'test.vue',
-    ast: descriptor.template!.ast,
-    source: descriptor.template!.content,
-    ssr: true,
-  })
-  expect(code).not.toMatch(`_ctx.t`)
+  expect(() =>
+    compileTemplate({
+      id: 'xxx',
+      filename: 'test.vue',
+      ast: descriptor.template!.ast,
+      source: descriptor.template!.content,
+      ssr: true,
+    }),
+  ).not.toThrowError()
 })
 
 interface Pos {

--- a/packages/compiler-sfc/__tests__/compileTemplate.spec.ts
+++ b/packages/compiler-sfc/__tests__/compileTemplate.spec.ts
@@ -427,6 +427,30 @@ test('prefixing edge case for reused AST', () => {
   expect(code).not.toMatch(`_ctx.t`)
 })
 
+test('prefixing edge case for reused AST ssr mode', () => {
+  const src = `
+  <script setup lang="ts">
+    import { Foo } from './foo'
+  </script>
+  <template>
+    <Bar>
+      <template #option="{ foo }"></template>
+    </Bar>
+  </template>
+  `
+  const { descriptor } = parse(src)
+  // compileScript triggers importUsageCheck
+  compileScript(descriptor, { id: 'xxx' })
+  const { code } = compileTemplate({
+    id: 'xxx',
+    filename: 'test.vue',
+    ast: descriptor.template!.ast,
+    source: descriptor.template!.content,
+    ssr: true,
+  })
+  expect(code).not.toMatch(`_ctx.t`)
+})
+
 interface Pos {
   line: number
   column: number

--- a/packages/compiler-ssr/src/transforms/ssrTransformComponent.ts
+++ b/packages/compiler-ssr/src/transforms/ssrTransformComponent.ts
@@ -55,7 +55,7 @@ import {
   ssrProcessTransitionGroup,
   ssrTransformTransitionGroup,
 } from './ssrTransformTransitionGroup'
-import { extend, isArray, isObject, isSymbol } from '@vue/shared'
+import { extend, isArray, isObject, isPlainObject, isSymbol } from '@vue/shared'
 import { buildSSRProps } from './ssrTransformElement'
 import {
   ssrProcessTransition,
@@ -120,6 +120,8 @@ export const ssrTransformComponent: NodeTransform = (node, context) => {
   // The branch is retrieved when processing slots again in ssr mode.
   const vnodeBranches: ReturnStatement[] = []
   const clonedNode = clone(node)
+
+  console.log(clonedNode)
 
   return function ssrPostTransformComponent() {
     // Using the cloned node, build the normal VNode-based branches (for
@@ -371,26 +373,12 @@ function subTransform(
 function clone(v: any): any {
   if (isArray(v)) {
     return v.map(clone)
-  } else if (isObject(v)) {
-    if (v instanceof Map) {
-      const res = new Map()
-      v.forEach((value: any, key: any) => {
-        res.set(clone(key), clone(value))
-      })
-      return res
-    } else if (v instanceof Set) {
-      const res = new Set()
-      v.forEach((value: any) => {
-        res.add(clone(value))
-      })
-      return res
-    } else {
-      const res: any = {}
-      for (const key in v) {
-        res[key] = clone(v[key])
-      }
-      return res
+  } else if (isPlainObject(v)) {
+    const res: any = {}
+    for (const key in v) {
+      res[key] = clone(v[key as keyof typeof v])
     }
+    return res
   } else {
     return v
   }

--- a/packages/compiler-ssr/src/transforms/ssrTransformComponent.ts
+++ b/packages/compiler-ssr/src/transforms/ssrTransformComponent.ts
@@ -372,11 +372,25 @@ function clone(v: any): any {
   if (isArray(v)) {
     return v.map(clone)
   } else if (isObject(v)) {
-    const res: any = {}
-    for (const key in v) {
-      res[key] = clone(v[key])
+    if (v instanceof Map) {
+      const res = new Map()
+      v.forEach((value: any, key: any) => {
+        res.set(clone(key), clone(value))
+      })
+      return res
+    } else if (v instanceof Set) {
+      const res = new Set()
+      v.forEach((value: any) => {
+        res.add(clone(value))
+      })
+      return res
+    } else {
+      const res: any = {}
+      for (const key in v) {
+        res[key] = clone(v[key])
+      }
+      return res
     }
-    return res
   } else {
     return v
   }

